llvm/lib/MC/MCSubtargetInfo.cpp

```diff
@@ -317,14 +317,20 @@ FeatureBitset MCSubtargetInfo::ApplyFeatureFlag(StringRef FS) {
 
 bool MCSubtargetInfo::checkFeatures(StringRef FS) const {
   SubtargetFeatures T(FS);
-  FeatureBitset Set, All;
-  for (std::string F : T.getFeatures()) {
-    ::ApplyFeatureFlag(Set, F, ProcFeatures);
-    if (F[0] == '-')
-      F[0] = '+';
-    ::ApplyFeatureFlag(All, F, ProcFeatures);
-  }
-  return (FeatureBits & All) == Set;
+  return all_of(T.getFeatures(), [this](const std::string &F) {
+    assert(SubtargetFeatures::hasFlag(F) &&
+           "Feature flags should start with '+' or '-'");
+    const SubtargetFeatureKV *FeatureEntry =
+        Find(SubtargetFeatures::StripFlag(F), ProcFeatures);
+    if (!FeatureEntry) {
+      errs() << "'" << F << "' is not a recognized feature for this target"
+             << " (ignoring feature)\n";
```

Comment:
Is ApplyFeatureFlag dead now? Or is it used during the construction of the bitset? Can the warning be left for that? We should migrate this to using DiagnosticInfo at least

Reply:
ApplyFeatureFlag is still used during construction.

I've changed this code to report a fatal error on invalid feature now. I don't think there's a good reason to make this a warning in checkFeatures().